pkg/pdfcpu/booklet.go

```diff
@@ -0,0 +1,314 @@
+	Margin    int         // Cropbox for booklet content within sheet
+	InpUnit   DisplayUnit // input display unit.
+}
+
+// BookletConfig returns an Booklet configuration
+func BookletConfig(desc string) (*Booklet, error) {
+	b := DefaultBookletConfig()
+	err := ParseBookletDetails(desc, b)
+	return b, err
+}
+
+// DefaultBookletConfig returns the default NUp configuration.
+func DefaultBookletConfig() *Booklet {
+	return &Booklet{
+		PageSize:  "A5",
+		SheetSize: "A3",
+		Margin:    0,
+	}
+}
+
+func (b Booklet) String() string {
+	return fmt.Sprintf("Booklet conf: input=%s %s, output=%s %s\n",
+		b.PageSize, *b.PageDim, b.SheetSize, b.SheetDim)
+}
+
+// ParseBookletDetails parses a Booklet command string into an internal structure.
+func ParseBookletDetails(s string, booklet *Booklet) error {
+	err1 := errInvalidNUpConfig
+	if s == "" {
+		return err1
+	}
+
+	ss := strings.Split(s, ",")
+
+	for _, s := range ss {
+
+		ss1 := strings.Split(s, ":")
+		if len(ss1) != 2 {
+			return err1
+		}
+
+		paramPrefix := strings.TrimSpace(ss1[0])
+		paramValueStr := strings.TrimSpace(ss1[1])
+
+		if err := bookletParamMap.Handle(paramPrefix, paramValueStr, booklet); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type bkltParamMap map[string]func(string, *Booklet) error
+
+// Handle applies parameter completion and if successful
+// parses the parameter values into import.
+func (m bkltParamMap) Handle(paramPrefix, paramValueStr string, booklet *Booklet) error {
+	var param string
+
+	// Completion support
+	for k := range m {
+		if !strings.HasPrefix(k, paramPrefix) {
+			continue
+		}
+		if len(param) > 0 {
+			return errors.Errorf("pdfcpu: ambiguous parameter prefix \"%s\"", paramPrefix)
+		}
+		param = k
+	}
+
+	if param == "" {
+		return errors.Errorf("pdfcpu: ambiguous parameter prefix \"%s\"", paramPrefix)
+	}
+
+	return m[param](paramValueStr, booklet)
+}
+
+var bookletParamMap = bkltParamMap{
+	"pagesize":  parseBookletPageSize,
+	"sheetsize": parseBookletSheetSize,
+	"margin":    parseBookletMargin,
+}
+
+func parseBookletPageSize(s string, b *Booklet) (err error) {
+	b.PageDim, b.PageSize, err = parsePageFormat(s)
+	return err
+}
+
+func parseBookletSheetSize(s string, b *Booklet) (err error) {
+	b.SheetDim, b.SheetSize, err = parsePageFormat(s)
+	return err
+}
+
+func parseBookletMargin(s string, b *Booklet) error {
+	f, err := strconv.ParseFloat(s, 64)
+	if err != nil {
+		return err
+	}
+
+	if f < 0 {
+		return errors.New("pdfcpu: booklet margin, Please provide a positive value")
+	}
+
+	b.Margin = int(toUserSpace(f, b.InpUnit))
+
+	return nil
+}
+
+// BookletFromPDF creates an booklet version of the PDF represented by xRefTable.
+func (ctx *Context) BookletFromPDF(selectedPages IntSet, booklet *Booklet) error {
+	grid := &Dim{
+		math.Round(booklet.SheetDim.Width / booklet.PageDim.Width),
+		math.Round(booklet.SheetDim.Height / booklet.PageDim.Height),
+	}
+	n := int(grid.Width * grid.Height)
+	if !(n == 2 || n == 4) {
```

Comment:
I think we should redesign the whole command to the way nup works.
`pdfcpu booklet [-p(ages) selectedPages] [description] outFile n inFile|imageFiles...``

Rather than asking the user to provide the input and output page size,
let him decide between 2up or 4up booklet (n= 2 or 4) and require a sheet page size (the size of the paper that gets printed). The user has to understand what that means so e have to explain it in the usage correspondingly. This way we can also reuse the Nup struct.

This also makes sense if you want to support booklets based on image files.

Reply:
Makes sense - will restructure a bit.